File: src/domain_ts/Color.ts
type RGBTuple = [number, number, number]

export enum ColorSpace {
  HSB,
  RGB
}

type Conversion = (
  v: number,
  p: number,
  q: number,
  t: number,
) => Color

type ConversionTable = {
  [key: number]: Conversion
}

const ConversionTableHSB: ConversionTable = {
  [0]: (v, p, q, t) => new Color(v, t, p),
  [1]: (v, p, q, t) => new Color(q, v, p),
  [2]: (v, p, q, t) => new Color(p, v, t),
  [3]: (v, p, q, t) => new Color(p, q, v),
  [4]: (v, p, q, t) => new Color(t, p, v),
  [5]: (v, p, q, t) => new Color(v, p, q),
  [6]: (v, p, q, t) => new Color(0, 0, 0),
}

export class Color {
  constructor (private red: number, private green: number, private blue: number) {
  }

  getR() { return Math.round(this.red * 255) }
  getG() { return Math.round(this.green * 255) }
  getB() { return Math.round(this.blue * 255) }

  static colorBetween (c1: Color, c2: Color, ratio: number): Color {
    const a = c1.red + ratio * (c2.red - c1.red);
    const b = c1.green + ratio * (c2.green - c1.green);
    const c = c1.blue + ratio * (c2.blue - c1.blue);

    return new Color(a, b, c);
  }

  static fromRGB (red: number, green: number, blue: number) {
    return new Color(
      red / 255,
      green / 255,
      blue / 255,
    );
  }

  static fromHSB (normalizedHue: number, normalizedSaturation: number, normalizedValue: number): Color {
    const region = Math.floor(normalizedHue * 6);
    const regionDecimal = (normalizedHue * 6) - region;
    const p = normalizedValue * (1 - normalizedSaturation);
    const q = normalizedValue * (1 - regionDecimal * normalizedSaturation);
    const t = normalizedValue * (1 - (1 - regionDecimal) * normalizedSaturation);
    
    return ConversionTableHSB[region](normalizedValue, p, q, t)
  }

  public intoRGBTuple (): RGBTuple {
    return [
      this.getR(),
      this.getG(),
      this.getB()
    ]
  }
}

File: src/MandelbrotSet/infrastructure/MandelbrotSetWasm/rust/mandelbrot_set/palette/color.rs
pub struct Color {
    r: f64,
    g: f64,
    b: f64,
}

impl Color {
    pub fn new(r: f64, g: f64, b: f64) -> Color {
        Color {
            r,
            g,
            b,
        }
    }

    pub fn color_between (c1: &Color, c2: &Color, ratio: f64) -> Color {
        let a = c1.r + ratio * (c2.r - c1.r);
        let b = c1.g + ratio * (c2.g - c1.g);
        let c = c1.b + ratio * (c2.b - c1.b);

        return Color::new(a, b, c);
    }


    pub fn get_r(&self) -> u8 {
        (self.r * 255.0).round() as u8
    }

    pub fn get_g(&self) -> u8 {
        (self.g * 255.0).round() as u8
    }

    pub fn get_b(&self) -> u8 {
        (self.b * 255.0).round() as u8
    }

    pub fn get_alpha(&self) -> u8 {
        255
    }
}
